### PR TITLE
Promisify and support Node.js >=4

### PR DESCRIPTION
**Pull request by kevva**
Cleaned up a lot of `vinyl` stuff and made it one URL only. With promises, one could do this now instead if they want to download multiple files at once:

```js
Promise.all(['foo.com/fish.jpg', 'bar.com/cat.png'].map(x => download(x, 'dest'))).then(() => {
    console.log('done!');
});
```

@sindresorhus @shinnn

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,211 +1,73 @@
...
+const Decompress = require('decompress');
+const filenamify = require('filenamify');
+const getStream = require('get-stream');
+const got = require('got');
+const mkdirp = require('mkdirp');
+const pify = require('pify');
+const fsP = pify(fs);
+
+const createPromise = (uri, output, stream, opts) => new Promise((resolve, reject) => {
+	stream.on('response', res => {
```
Comment (SamVerschueren) [inline index.js]: If you really want to go into promises, you could use [rfpify](https://github.com/samverschueren/rfpify) here. But definitely not a necessarity :).

File: index.js [hunk 2]
```diff
@@ -1,211 +1,73 @@
...
- *
- * @param {Function|String} name
- * @api public
- */
-
-Download.prototype.rename = function (name) {
-	if (!arguments.length) {
-		return this._rename;
+	if (opts.extract) {
+		return new Promise((resolve, reject) => {
```
Comment (SamVerschueren) [inline index.js]: Can be written like this.

```js
const decompress = new Decompress(opts).src(data).dest(path.dirname(output));
return pify(decompress.run.bind(decompress))();
```

**Comment by SamVerschueren** [comment]
Nice

File: index.js (hunk 2 above)
Comment (kevva) [inline index.js]: Ohh, much better :).

File: index.js (hunk 1 above)
Comment (kevva) [inline index.js]: If I use `rfpify`, how do I return both the `error` listener and the `response`?

Comment (SamVerschueren) [inline index.js]: After looking into it again, I don't think there is that much benefit. You still need the wrapper promise.

```js
rfpify(stream.on.bind(stream)('response')
    .then(res => opts.encoding === null ? getStream.buffer(res) : getStream(res, opts))
    .then(resolve)
    .catch(reject);

stream.on('error', reject);
```

Comment (kevva) [inline index.js]: Yup, I'll just leave it as is.

File: index.js (hunk 2 above)
Comment (kevva) [inline index.js]: Next up: rewrite `decompress` :(.

File: readme.md [hunk 3]
```diff
@@ -14,109 +14,50 @@ $ npm install --save download
...
 
-Type: `number`  
-Default: `0`
+## API
 
-Remove leading directory components from extracted files.
+### download(url, [destination], [options])
 
-### .get(url, [dest])
+Returns both a Promise for a buffer and a Duplex stream with [additional events](https://github.com/sindresorhus/got#streams).
```
Comment (sindresorhus) [inline readme.md]: buffer => Buffer

Comment (sindresorhus) [inline readme.md]: I would also linkify `Duplex stream` to the Node.js docs, as most people wouldn't know how it works.

File: readme.md [hunk 4]
```diff
@@ -14,109 +14,50 @@ $ npm install --save download
...
-Default: `false`
-
-If set to `true`, try extracting the file using [decompress](https://github.com/kevva/decompress/).
-
-##### options.mode
-
-Type: `string`
-
-Set mode on the downloaded file, i.e `{mode: '755'}`.
+download('http://unicorn.com/foo.jpg').pipe(fs.createWriteStream('dist/foo.jpg'));
```
Comment (sindresorhus) [inline readme.md]: Maybe show that you can drop the `http://` prefix.

File: index.js [hunk 5]
```diff
@@ -1,211 +1,62 @@
...
+const pify = require('pify');
+const fsP = pify(fs);
+
+const createPromise = (uri, output, stream, opts) => new Promise((resolve, reject) => {
+	stream.on('response', res => {
+		const stream = opts.encoding === null ? getStream.buffer(res) : getStream(res, opts);
+
+		stream
+			.then(data => resolve(data))
+			.catch(err => reject(err));
```
Comment (sindresorhus) [inline index.js]:
```js
stream.then(resolve).catch(reject);
```

File: index.js [hunk 6]
```diff
@@ -1,211 +1,62 @@
...
- * @param {Function|String} name
- * @api public
- */
-
-Download.prototype.rename = function (name) {
-	if (!arguments.length) {
-		return this._rename;
+	if (opts.extract) {
+		const decompress = new Decompress(opts).src(data).dest(path.dirname(output));
+		return pify(decompress.run.bind(decompress))().then(() => data);
```
Comment (sindresorhus) [inline index.js]: I assume the next step is to rewrite `decompress` :p

File: index.js [hunk 7]
```diff
@@ -1,211 +1,62 @@
...
-};
-
-/**
- * Run
- *
- * @param {Function} cb
- * @api public
- */
+	return pify(mkdirp)(path.dirname(output))
+		.then(() => fsP.writeFile(output, data))
```
Comment (sindresorhus) [inline index.js]: Not sure whether this is clear enough, but the last expression is returned, so you could just append `data`:

```js
.then(() => fsP.writeFile(output, data), data)
```

**Comment by sindresorhus** [comment]
> With promises, one could do this now instead if they want to download multiple files at once:

I would show an example of this, like the one in the description.

File: index.js (hunk 7 above)
Comment (SamVerschueren) [inline index.js]: I think I rather prefer the more expressive version.

**Comment by sindresorhus** [comment]
Good cleanup :)

File: index.js (hunk 6 above)
Comment (kevva) [inline index.js]: Already in progress :).

File: index.js (hunk 7 above)
Comment (kevva) [inline index.js]: Yeah, me too. Makes it easier to read which is more important imo.

**Comment by kevva** [comment]
@SamVerschueren @sindresorhus, thanks for reviewing :). Going to merge this now.

File: index.js (hunk 6 above)
Comment (kevva) [inline index.js]: Woho https://github.com/kevva/download/commit/7f8ab78e926301b92923a580320cbbd0fc7a5584.

**Pull request merged**